src/Interpreters/Cache/FileCache.cpp

```diff
@@ -136,7 +140,18 @@ const FileCache::UserInfo & FileCache::getInternalUser()
 
 bool FileCache::isInitialized() const
 {
-    return is_initialized.load(std::memory_order_seq_cst);
+    if (load_metadata_asynchronously)
+    {
+        return is_initialized;
+    }
+    else
+    {
+        if (init_exception)
+        {
+            std::rethrow_exception(init_exception);
```

Comment:
@kssenii I tried to read previous comments but probably did not understand them.
Why do we want to change behaviour here ? 
The `FileCache` interface was previously designed naturally and can be used for async load and sync load.
Client before trying to use cache check if cache initialized using `isInitialized` method and if it returns false, fallback to reading method without cache. If client accidentally did not check `isInitialized` and tried to use cache, any `FileCache` method throws exception using `assertInitialized`. 
Throwing exception in `isInitialized` method does not look very obvious, because no one can expect that. What is the motivation ? 
Maybe for client it is better to manually use `assertInitialized` if he wants to get Exception.

Reply:
Previously, before this PR, we would try to read from cache without checking `isInitialized` and will directly go to `assertInitialized`, which was the expectation that cache must always be initialized on server startup. But now in this PR we would not try to use cache for read-through/write-through cache if `isInitialized` returns `false` (because of the new check in ReadBufferFromRemoteFSGather). But it will return `false` in case initialization failed as well. So the previous behaviour changes - that with non-async cache initialization we will no longer catch that cache failed to be initialized. I want to preserve previous default behaviour, which is possible without throwing exception from `isInitialized` though, I was not forcing exactly this, but just preserving old behaviour.

Reply:
I've moved the logic to throw to a separate method `throwInitExceptionIfNeeded`